Commonalities/documentation/API-design-guidelines.md

```diff
@@ -1021,3 +1027,245 @@ Finally, this part describes the OAuth security applied to the API. This spec is
 - Endpoint token URL
 - URL to endpoint authorization ( If flow is based on "`authorizationCode`").	
 
+
+## 12. Subscription, Notification & Event
+
+In order to provide event-based interaction, CAMARA API could provided capabilities for subscription & notification management.
+A subscription allows to an API consumer to request event notification reception at a given url (callback address) for a specific context.
+A notification is the publication at the listener address about an occurred event.
+Event type managed are explicitly defined in CAMAPA API OAS.
+
+### 12.1 Subscription
+
+We distinguish 2 types of subscription:
+- Instance-based subscription (indirect creation)
+- Resource-based subscription (direct creation)
+
+**Instance-based subscription**
+
+An instance-based subscription is a subscription indirectly created, additionally, to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.
+
+Providing this capability is optional for any CAMARA APIs depending on UC requirements
+
+If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:
+
+| attribute name | type | attribute description | cardinality |
+| ----- |	-----  |	 -----  | -----  | 
+| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |
+| notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |
```

Comment:
@eric-murray 

The summary from your side seems to be, from an API provider perspective, `notificationAuthToken` is a string that may or may not be an OAuth2 token and it is upto the client (API consumer's) implementation. API provider has to just play it back.
@shilpa-padgaonkar @PedroDiez @bigludo7 - is this your understanding as well?

My concern is that we are making it ambiguous and encouraging misuse by specifying them in guidelines.
* If an OAuth2 access token is indeed used here, this is a non-standard practice and usage of OAuth2 access tokens  (even opaque). We could avoid this misuse by making this `notificationAuthToken` as part of message body instead of `Authorization: Bearer` header (or) we should clearly state it should not be an OAuth2 access token. By stating this, CAMARA & the implementations will not become party to the non-standard usage of OAuth2 access tokens and it is only the operator that chooses to do it this way.
* If it cannot be OAuth2 access token and it is an opaque string that is used as correlation value, we should use it in the message body instead of `Authorization` header - again preventing non-standard usage being encouraged in the guidelines.

If we, as a community, still want to allow the possibility of OAuth2 opaque access token being used in a non-standard way, the guidelines should allow implementations to ignore this field even when this optional parameter is present.

Reply:
> The summary from your side seems to be, from an API provider perspective, notificationAuthToken is a string that may or may not be an OAuth2 token and it is upto the client (API consumer's) implementation. API provider has to just play it back.

Yes, also my understanding

Reply:
Indeed, from my understanding, in case an API client decides to use OAuth2 Token it is based on its infra, so API Provider is not involved in its generation.

For the other point of view you comment @sfnuser, within CAMARA is already defined the concept of "x-correlator" as a header parameter

Reply:
@bigludo7 @PedroDiez 
Thanks for the update.

I want to reiterate my concern that I posted [here](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1180504699). As per [IANA](https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml), the `Bearer` type in `Authorization: Bearer` header means it is an OAuth2 token. However, we state that it is upto the API user implementation (i.e.) it may or may not be an OAuth2 token. The guidelines also makes the API implementations to receive this token in message body and replay it back in header - again a non-standard usage.

@shilpa-padgaonkar - Is this proposal inline with the `Security` section of the guidelines?

Alternatively, we could consider explicitly supporting both options - OAuth2 (notificationAuthToken) or a CorrelationId. For e.g. the API user can provide `notificationUrl` and have to choose one of the three options mentioned below:
* None - The notification messages will be posted to the `notificationUrl` and the `event` contents are enough for the operator. For some cases, this is perfectly acceptable.
* CorrelationId - The API user should POST this ID in the message `body` when they register for notification with the API provider. The API provider will post this `correlationId` back in the message `body` along with the `event` contents
* OAuth2 Access Token - The API user should POST `notificationUrl`, `tokenUrl` & `clientId, clientSecret` in some hash format (assuming `clientCredentials` flow suits all APIs notification callback) during notification registration and the API provider will use these as per the `standard` OAuth2 framework when posting notifications. The API server may also choose to reject this option in the registration response if it is too much resources and propose alternatives in the message body (None or CorrelationId)

I haven't fully thought through all the cases but worth pondering the pros & cons rather than sticking with the non-standard way.

As I requested earlier, if we still want to go ahead as it is, CAMARA should not enforce non-standard ways in implementations & allow it to ignore this field even when this optional `notificationAuthToken` field is present.